webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala

```diff
@@ -1181,127 +1193,138 @@ class ComplianceAPIService(
+                                  case (_, (rule, _, _)) => RuleTarget.merge(rule.targets).includes(nodeGroupsInfo(g)._3)
+                                }
+
+                                (g, GlobalTargetInfo(name, serverList, globalRulesByGroup, targetedRulesByGroup))
+                            }
+
+        level = Some(1)
+
+        bothGlobalTargeted <-
+          ZIO.foreach(globalTargetInfos) {
+            case (id, info) =>
+              (
+                getByNodeGroupCompliance(
+                  id,
+                  info.name,
+                  info.nodeIds,
+                  directiveLib.allDirectives,
+                  nodeFacts,
+                  nodeSettings,
+                  info.globalRulesByGroup,
+                  level,
+                  isGlobalCompliance = true
+                ) <&> getByNodeGroupCompliance(
+                  id,
+                  info.name,
+                  info.nodeIds,
+                  directiveLib.allDirectives,
+                  nodeFacts,
+                  nodeSettings,
+                  info.targetedRulesByGroup,
+                  level,
+                  isGlobalCompliance = false
+                )
+              ).map(
+                (id, _)
               )
-            }
-            val policyMode      =
-              getRulePolicyMode(r, directiveLib.allDirectives, targetedNodeIds.toSet, nodeSettings, globalPolicyMode)
-
-            (r.id, (targetedNodeIds, policyMode))
-          })
-          .toMap
+          }
+      } yield {
+        bothGlobalTargeted.toMap
       }
-
-      // global compliance : filter our rules that are applicable to any node in this group
-      globalRulesByGroup   = nodeGroupsInfo.map {
-                               case (g, (_, serverList, _)) =>
-                                 (
-                                   g,
-                                   rules.filter(rule => {
-                                     val targetedNodeIds = allRuleInfos(rule.id)._1
-                                     targetedNodeIds.exists(serverList.contains(_))
-                                   })
-                                 )
-                             }
-      // targeted compliance : filter rules that only include this group in its targets
-      targetedRulesByGroup = globalRulesByGroup.map {
-                               case (g, rules) =>
-                                 (g, rules.filter(rule => RuleTarget.merge(rule.targets).includes(nodeGroupsInfo(g)._3)))
-                             }
-
-      level = Some(1)
-
-      bothGlobalTargeted <-
-        ZIO.foreach(nodeGroupsInfo.toList) {
-          case (id, (name, serverList, _)) =>
-            (getByNodeGroupCompliance(
-              id,
-              name,
-              serverList,
-              directiveLib.allDirectives,
-              nodeFacts,
-              nodeSettings,
-              globalRulesByGroup(id),
```

Comment:
we really want to avoid that. It is not very resilient to refactoring. 
Here, we avoid it again by making is unrepresentable with the datastructure